lib/magma/server/update.rb

```diff
@@ -55,4 +60,67 @@ def load_revisions
     log(m.complaints)
     @errors.concat(m.complaints)
   end
+
+  def update_any_file_links
+    # Here, if there are any File attributes in the revisions,
+    #   we'll send a request to the Metis "copy" route, using the
+    #   Etna::Client.
+    # Assumes a conversion to Metis storage and no other
+    #   provider will be set.
+
+    # Etna::Client.new throws an exception if the host config includes protocol,
+    # i.e. https://metis-dev.etna-development.org will throw an exception
+    # raise Etna::BadRequest, 'Storage host should not include protocol' if
+    #   Magma.instance.config(:storage).fetch(:host).start_with? 'http'
+
+    @revisions.each do |model, model_revisions|
+      model_revisions.each do |revision|
+
+        # The below test for :stats as a File indicator seems
+        #   brittle -- anything better?
+        if revision.to_loader.key?(:stats)
+          auth = Etna::Auth.new(:magma)
+
+          host = Magma.instance.config(:storage).fetch(:host)
+
+          client = Etna::Client.new(
+            "https://#{host}",
+            @request.cookies[Magma.instance.config(:token_name)] || auth.send('auth', *[@request, :etna]))
+
+          copy_route = ''
+
+          client.routes.each do |route|
+            if route[:name] == 'copy'
+              copy_route = route
+              break
+            end
+          end
+
+          if copy_route == ''
+            next
+          end
+
+          path = client.send(
+            'route_path',
+            *[copy_route, {:project_name => @project_name, :bucket_name => "magma", :file_path => "test.txt"}])
+
+          # Now populate the standard headers
+          hmac_params = {
+            method: 'post',
+            host: host,
+            path: path,
+
+            expiration: (DateTime.now + 10).iso8601,
+            id: 'magma',
+            nonce: SecureRandom.hex,
+            headers: { project_name: @project_name, action: 'copy' },
```

Comment:
Leave headers as empty Hash.

Reply:
This is used to include extra parameters we want signed in HMAC signature, like user ID or something.

Reply:
We will have optional parameters that should get hashed, `new_bucket_name`, `new_file_name`, and maybe `new_project_name`

Reply:
`new_bucket_name == magma`, `new_file_name` == magma's controlled name for the file. We should generate that somehow? From FileAttribute to see how it gets done now when a file gets put into S3.